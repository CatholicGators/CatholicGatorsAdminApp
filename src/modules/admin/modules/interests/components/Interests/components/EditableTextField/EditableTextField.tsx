import React, { Component } from "react";
import classnames from "classnames";
import {
    Theme,
    createStyles,
    withStyles,
    IconButton,
    Input,
    Typography,
} from "@material-ui/core";
import { Delete, Edit, Save, Close } from "@material-ui/icons";

export const styles = (theme: Theme) =>
    createStyles({
        row: {
            display: "flex",
            alignItems: "center",
            width: "100%",
        },
        input: {
            margin: 0,
        },
        actions: {
            display: "none",
            marginLeft: `${theme.spacing(1)}px`,
        },
        editActions: {
            display: "flex",
            alignItems: "center",
        },
        show: {
            display: "block",
        },
    });

export type Props = {
    classes: any;
    isEditing: boolean;
    isHovered: boolean;
    id: string;
    text: string;
    beginEditing: (id: string) => void;
    cancelEditing: () => void;
    deleteText: (id: string) => void;
    save: (id: string, newText: string) => void;
};

type State = {
    editedText: string;
};

export class EditableTextField extends Component<Props, State> {
    state = {
        editedText: this.props.text,
    };

    onEditChange(newText: string) {
        this.setState({ editedText: newText });
    }

    render() {
        const {
            isEditing,
            isHovered,
            id,
            text,
            classes,
            beginEditing,
            deleteText,
            save,
            cancelEditing,
        } = this.props;

        const { editedText } = this.state;

        return isEditing ? (
            <div className={classes.row}>
                <Input
                    fullWidth
                    className={classes.input}
                    value={this.state.editedText}
                    onChange={(ev) => this.onEditChange(ev.target.value)}
                />
                <div className={classes.editActions}>
                    <IconButton id="save" onClick={() => save(id, editedText)}>
                        <Save />
                    </IconButton>
                    <IconButton id="cancel" onClick={() => cancelEditing()}>
                        <Close />
                    </IconButton>
                </div>
            </div>
        ) : (
            <div className={classes.row}>
                <Typography>{text}</Typography>
                <div
                    className={classnames({
                        [classes.actions]: true,
                        [classes.show]: isHovered,
                    })}
                >
                    <IconButton id="edit" onClick={() => beginEditing(id)}>
                        <Edit />
                    </IconButton>
                    <IconButton id="delete" onClick={() => deleteText(id)}>
                        <Delete />
                    </IconButton>
                </div>
            </div>
        );
    }
}

export default withStyles(styles)(EditableTextField);
