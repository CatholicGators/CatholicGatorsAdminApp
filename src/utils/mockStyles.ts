import { createMuiTheme } from '@material-ui/core'

export default (styles) => Object.keys(styles(createMuiTheme())).reduce((acc, curr) => ({...acc, [curr]: "test"}), {})