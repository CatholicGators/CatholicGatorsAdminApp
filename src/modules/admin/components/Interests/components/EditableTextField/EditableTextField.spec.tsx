import React from 'react'
import { shallow } from 'enzyme';

import { EditableTextField, Props, styles } from './EditableTextField'
import {
    Edit,
    Delete,
    Save,
    Close
} from '@material-ui/icons';
import { Input } from '@material-ui/core';
import mockStyles from '../../../../../../utils/mockStyles';

describe('EditableTextField', () => {
    let props: Props, wrapper

    beforeEach(() => {
        props = {
            classes: mockStyles(styles),
            isEditing: false,
            isHovered: false,
            id: "1",
            text: "totally testing",
            beginEditing: jest.fn(),
            cancelEditing: jest.fn(),
            deleteText: jest.fn(),
            save: jest.fn()
        }
        wrapper = shallow(<EditableTextField {...props} />)
    })

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('defaults the editedText to the text of the option that is passed in', () => {
        expect(wrapper.state().editedText).toBe(props.text)
    })

    describe('when not editing', () => {
        it('renders the edit and delete buttons', () => {
            expect(wrapper.exists(Edit)).toBe(true)
            expect(wrapper.exists(Delete)).toBe(true)
            expect(wrapper.exists(Save)).toBe(false)
            expect(wrapper.exists(Close)).toBe(false)
            expect(wrapper.exists(Input)).toBe(false)
        })

        it('beginEdit is called when edit clicked', () => {
            const edit = wrapper.find('#edit')
            edit.simulate('click')

            expect(props.beginEditing).toHaveBeenCalledWith(props.id)
        })

        it('cancelEdit is called when cancel clicked', () => {
            const deleteButton = wrapper.find('#delete')
            deleteButton.simulate('click')

            expect(props.deleteText).toHaveBeenCalledWith(props.id)
        })
    })

    describe('when editing', () => {
        beforeEach(() => {
            wrapper.setProps({ isEditing: true })
        })

        it('renders the input, save, and cancel buttons', () => {
            expect(wrapper.exists(Edit)).toBe(false)
            expect(wrapper.exists(Delete)).toBe(false)
            expect(wrapper.exists(Save)).toBe(true)
            expect(wrapper.exists(Close)).toBe(true)
            expect(wrapper.exists(Input)).toBe(true)
        })

        it('save is called when save clicked', () => {
            const newText = 'test'
            const input = wrapper.find(Input)
            input.simulate("change", { target: { value: newText } });
            const save = wrapper.find('#save')
            save.simulate('click')

            expect(props.save).toHaveBeenCalledWith(props.id, newText)
        })

        it('cancelEdit is called when cancel clicked', () => {
            const cancel = wrapper.find('#cancel')
            cancel.simulate('click')

            expect(props.cancelEditing).toHaveBeenCalled()
        })
    })
})
