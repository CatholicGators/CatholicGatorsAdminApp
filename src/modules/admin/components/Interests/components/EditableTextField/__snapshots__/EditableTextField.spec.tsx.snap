// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableTextField should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditableTextField
    beginEditing={[MockFunction]}
    cancelEditing={[MockFunction]}
    classes={Object {}}
    deleteText={[MockFunction]}
    id={1}
    isEditing={false}
    isHovered={false}
    save={[MockFunction]}
    text="totally testing"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <WithStyles(Typography)>
          totally testing
        </WithStyles(Typography)>,
        <div
          className=""
        >
          <WithStyles(IconButton)
            id="edit"
            onClick={[Function]}
          >
            <pure(EditIcon) />
          </WithStyles(IconButton)>
          <WithStyles(IconButton)
            id="delete"
            onClick={[Function]}
          >
            <pure(DeleteIcon) />
          </WithStyles(IconButton)>
        </div>,
      ],
      "className": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "totally testing",
        },
        "ref": null,
        "rendered": "totally testing",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <WithStyles(IconButton)
              id="edit"
              onClick={[Function]}
            >
              <pure(EditIcon) />
            </WithStyles(IconButton)>,
            <WithStyles(IconButton)
              id="delete"
              onClick={[Function]}
            >
              <pure(DeleteIcon) />
            </WithStyles(IconButton)>,
          ],
          "className": "",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <pure(EditIcon) />,
              "id": "edit",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <pure(DeleteIcon) />,
              "id": "delete",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <WithStyles(Typography)>
            totally testing
          </WithStyles(Typography)>,
          <div
            className=""
          >
            <WithStyles(IconButton)
              id="edit"
              onClick={[Function]}
            >
              <pure(EditIcon) />
            </WithStyles(IconButton)>
            <WithStyles(IconButton)
              id="delete"
              onClick={[Function]}
            >
              <pure(DeleteIcon) />
            </WithStyles(IconButton)>
          </div>,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "totally testing",
          },
          "ref": null,
          "rendered": "totally testing",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <WithStyles(IconButton)
                id="edit"
                onClick={[Function]}
              >
                <pure(EditIcon) />
              </WithStyles(IconButton)>,
              <WithStyles(IconButton)
                id="delete"
                onClick={[Function]}
              >
                <pure(DeleteIcon) />
              </WithStyles(IconButton)>,
            ],
            "className": "",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <pure(EditIcon) />,
                "id": "edit",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <pure(DeleteIcon) />,
                "id": "delete",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
